packages/table/src/interactions/reorderable.tsx

```diff
@@ -0,0 +1,154 @@
+import { Draggable, ICoordinateData, IDraggableProps } from "../interactions/draggable";
+import { IRegion, RegionCardinality, Regions } from "../regions";
+
+export interface IReorderableProps {
+    /**
+     * A callback that is called while the user is dragging to reorder.
+     *
+     * @param oldIndex the original index of the element or set of elements
+     * @param newIndex the new index of the element or set of elements
+     * @param length the number of contiguous elements that were moved
+     */
+    onReorderPreview: (oldIndex: number, newIndex: number, length: number) => void;
+
+    /**
+     * A callback that is called when the user is done dragging to reorder.
+     *
+     * @param oldIndex the original index of the element or set of elements
+     * @param newIndex the new index of the element or set of elements
+     * @param length the number of contiguous elements that were moved
+     */
+    onReorder: (oldIndex: number, newIndex: number, length: number) => void;
+
+    /**
+     * When the user reorders something, this callback is called with a new
+     * array of `Region`s. This array should be considered the new selection
+     * state for the entire table.
+     */
+    onSelection: (regions: IRegion[]) => void;
+
+    /**
+     * An array containing the table's selection Regions.
+     */
+    selectedRegions: IRegion[];
+}
+
+export interface IDragReorderable extends IReorderableProps {
+    /**
+     * A callback that determines a `Region` for the single `MouseEvent`. If
+     * no valid region can be found, `null` may be returned.
+     */
+    locateClick: (event: MouseEvent) => IRegion;
+
+    /**
+     * A callback that determines the index at which to show the preview guide.
+     * This is equivalent to the absolute index in the old ordering where the
+     * reordered element will move.
+     */
+    locateDrag: (event: MouseEvent, coords: ICoordinateData) => number;
+
+    /**
+     * A callback that converts the provided index into a region. The returned
+     * region will be used to update the current selection after drag-reordering.
+     */
+    toRegion: (index1: number, index2?: number) => IRegion;
+}
+
+@PureRender
+export class DragReorderable extends React.Component<IDragReorderable, {}> {
+    public static isLeftClick(event: MouseEvent) {
+        return event.button === 0;
+    }
+
+    private selectedRegionStartIndex: number;
+    private selectedRegionLength: number;
+
+    public render() {
+        const draggableProps = this.getDraggableProps();
+        return (
+            <Draggable {...draggableProps} preventDefault={false}>
+                {this.props.children}
+            </Draggable>
+        );
+    }
+
+    private getDraggableProps(): IDraggableProps {
+        return this.props.onReorder == null ? {} : {
+            onActivate: this.handleActivate,
+            onDragEnd: this.handleDragEnd,
+            onDragMove: this.handleDragMove,
+        };
+    }
```

Comment:
there seems to be a bit of duplication between this and `<DragSelectable>` i'm not sure how to de-dupe, but it seems like we could